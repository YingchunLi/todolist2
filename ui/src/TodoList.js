import React, {Component} from 'react';
import PropTypes from 'prop-types';

const TodoListTable = ({title, todos}) =>
    <table className="table table-striped table-hover">
      <caption>{title}</caption>
      <thead>
        <tr>
          <th>name</th>
          <th>description</th>
          <th>due date</th>
          <th>status</th>
        </tr>
      </thead>
      <tbody>
        {
          todos.map((todo, idx) =>
            <tr key={idx} >
              <td>{todo.name}</td>
              <td>
                {todo.description &&
                todo.description.split('\n')
                  .map((line,idx) => <div key={idx}>{line}</div>)
                }
                </td>
              <td>{todo.dueDate && new Date(todo.dueDate).toISOString()}</td>
              <td>{todo.status}</td>
            </tr>
          )
        }
      </tbody>
    </table>;

export class TodoList extends Component {
  state = {
    todos: [],
  };

  componentWillMount() {
    fetch('/api/todos').then(response => response.json())
      .then(json => this.setState({todos: json}));
  }

  render() {
    const {todos} = this.state;
    const pendingTodos = todos.filter(t => t.status === 'Pending');
    const doneTodos = todos.filter(t => t.status === 'Done');
    return (
      <div>
        <h2>Todo List</h2>
        <TodoListTable title="Pending todos" todos={pendingTodos} />
        <TodoListTable title="Done todos" todos={doneTodos} />
      </div>
    );
  }
}

TodoList.propTypes = {
  todos:  PropTypes.arrayOf(PropTypes.object),
};

TodoList.defaultProps = {
  todos: [],
};

export default TodoList;
