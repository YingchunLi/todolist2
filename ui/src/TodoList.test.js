import React from 'react';
import ReactDOM from 'react-dom';
import renderer from "react-test-renderer";

import TodoList from './TodoList';

it('renders without todos', () => {
  fetch.once('[]');
  const div = document.createElement('div');
  ReactDOM.render(<TodoList />, div);
});

it('renders with todos', () => {
  const body = [{
    id:1,
    name:'milk',
    description: 'remember the milk',
    dueDate: '2018-12-31T00:00:00.000z',
    status: 'Pending'}
  ];
  fetch.once( JSON.stringify(body) );
  const div = document.createElement('div');
  ReactDOM.render(<TodoList />, div);
});

it('snapshot test - without none-ok response', () => {
  fetch.once('some internal server error', {status: 500, statusText: 'Internal Server Error'});
  const component = renderer.create(
    <TodoList />
  ).toJSON();
  expect(component).toMatchSnapshot();
});

it('snapshot test - without none-json resonse', () => {
  fetch.once('some response not in json format');
  const component = renderer.create(
    <TodoList />
  ).toJSON();
  expect(component).toMatchSnapshot();
});

it('snapshot test - without todos', () => {
  fetch.once('[]');
  const component = renderer.create(
    <TodoList />
  ).toJSON();
  expect(component).toMatchSnapshot();
});

it('snapshot test - with todos', () => {
  const body = [{
    id:1,
    name:'milk',
    description: 'remember the milk',
    dueDate: '2018-12-31T00:00:00.000z',
    status: 'Pending'}
  ];
  fetch.once( JSON.stringify(body) );
  const component = renderer.create(
    <TodoList />
  ).toJSON();

  expect(component).toMatchSnapshot();
});