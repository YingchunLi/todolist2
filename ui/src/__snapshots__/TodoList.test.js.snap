// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot test - with todos 1`] = `
<div>
  <h2>
    Todo List
  </h2>
  <table
    className="table table-striped table-hover"
  >
    <caption>
      Pending todos
    </caption>
    <thead>
      <tr>
        <th>
          name
        </th>
        <th>
          description
        </th>
        <th>
          due date
        </th>
        <th>
          status
        </th>
      </tr>
    </thead>
    <tbody />
  </table>
  <table
    className="table table-striped table-hover"
  >
    <caption>
      Done todos
    </caption>
    <thead>
      <tr>
        <th>
          name
        </th>
        <th>
          description
        </th>
        <th>
          due date
        </th>
        <th>
          status
        </th>
      </tr>
    </thead>
    <tbody />
  </table>
</div>
`;

exports[`snapshot test - without none-json resonse 1`] = `
<div>
  <h2>
    Todo List
  </h2>
  <table
    className="table table-striped table-hover"
  >
    <caption>
      Pending todos
    </caption>
    <thead>
      <tr>
        <th>
          name
        </th>
        <th>
          description
        </th>
        <th>
          due date
        </th>
        <th>
          status
        </th>
      </tr>
    </thead>
    <tbody />
  </table>
  <table
    className="table table-striped table-hover"
  >
    <caption>
      Done todos
    </caption>
    <thead>
      <tr>
        <th>
          name
        </th>
        <th>
          description
        </th>
        <th>
          due date
        </th>
        <th>
          status
        </th>
      </tr>
    </thead>
    <tbody />
  </table>
</div>
`;

exports[`snapshot test - without none-ok response 1`] = `
<div>
  <h2>
    Todo List
  </h2>
  <table
    className="table table-striped table-hover"
  >
    <caption>
      Pending todos
    </caption>
    <thead>
      <tr>
        <th>
          name
        </th>
        <th>
          description
        </th>
        <th>
          due date
        </th>
        <th>
          status
        </th>
      </tr>
    </thead>
    <tbody />
  </table>
  <table
    className="table table-striped table-hover"
  >
    <caption>
      Done todos
    </caption>
    <thead>
      <tr>
        <th>
          name
        </th>
        <th>
          description
        </th>
        <th>
          due date
        </th>
        <th>
          status
        </th>
      </tr>
    </thead>
    <tbody />
  </table>
</div>
`;

exports[`snapshot test - without todos 1`] = `
<div>
  <h2>
    Todo List
  </h2>
  <table
    className="table table-striped table-hover"
  >
    <caption>
      Pending todos
    </caption>
    <thead>
      <tr>
        <th>
          name
        </th>
        <th>
          description
        </th>
        <th>
          due date
        </th>
        <th>
          status
        </th>
      </tr>
    </thead>
    <tbody />
  </table>
  <table
    className="table table-striped table-hover"
  >
    <caption>
      Done todos
    </caption>
    <thead>
      <tr>
        <th>
          name
        </th>
        <th>
          description
        </th>
        <th>
          due date
        </th>
        <th>
          status
        </th>
      </tr>
    </thead>
    <tbody />
  </table>
</div>
`;
